### PR TITLE
Move ES_TMPDIR substitution into jvm options parser

### LINKED ISSUE #47133: JvmOptionsParser - cannot resolve enviroment variables
Version: **ES 7.3.2**
OS: **CentOS Linux release 7.6.1810 (Core)**

Change the path to `gc.log` in `jvm.options`:
```
9-:-Xlog:gc*,gc+age=trace,safepoint:file=${ES_TMPDIR}/gc.log:utctime,pid,tags:filecount=32,filesize=64m
```

ES cannot start:
```
Exception in thread "main" java.lang.RuntimeException: starting java failed with [1]
output:
[0.000s][error][logging] Error opening log file '${ES_TMPDIR}/gc.log': No such file or directory
[0.000s][error][logging] Initialization of output 'file=${ES_TMPDIR}/gc.log' using options 'filecount=32,filesize=64m' failed.
error:
OpenJDK 64-Bit Server VM warning: Option UseConcMarkSweepGC was deprecated in version 9.0 and will likely be removed in a future release.
Invalid -Xlog option '-Xlog:gc*,gc+age=trace,safepoint:file=${ES_TMPDIR}/gc.log:utctime,pid,tags:filecount=32,filesize=64m', see error log for details.
Error: Could not create the Java Virtual Machine.
Error: A fatal exception has occurred. Program will exit.
        at org.elasticsearch.tools.launchers.JvmErgonomics.flagsFinal(JvmErgonomics.java:111)
        at org.elasticsearch.tools.launchers.JvmErgonomics.finalJvmOptions(JvmErgonomics.java:79)
        at org.elasticsearch.tools.launchers.JvmErgonomics.choose(JvmErgonomics.java:57)
        at org.elasticsearch.tools.launchers.JvmOptionsParser.main(JvmOptionsParser.java:89)
```

### PR DESCRIPTION
**Pull request by jasontedor**
This commit moves the ES_TMPDIR substitution that we do for JVM options into the JVM options parser itself. This solves a problem where the fact that the we do not make the substitution before ergonomics parsing can lead to the JVM that we start for computing the ergonomic values failing to start. Additionally, moving this substitution here enables us to simplify the shell scripts since we do not need to implement this there, and twice for Bash and Windows.

Closes #47133

### CODE REVIEW

**Comment by jasontedor** [comment]
@rjernst This is only a draft, I need to add tests, and actually test this on Windows (I haven't) so no need for a formal review, really looking for feedback on the direction I'm taking this.

**Comment by rjernst** [comment]
The direction looks good.

**Comment by jasontedor** [comment]
@rjernst This is ready for you.

**Comment by jasontedor** [comment]
@rjernst Can you review?

File: distribution/tools/launchers/src/main/java/org/elasticsearch/tools/launchers/JvmOptionsParser.java [hunk 1]
```diff
@@ -115,6 +117,24 @@ public void accept(final int lineNumber, final String line) {
...
+                   actualJvmOption = actualJvmOption.replace(placeholderSubstitution.getKey(), placeholderSubstitution.getValue());
+                }
+                substitutedJvmOptions.add(actualJvmOption);
+            } else {
+                substitutedJvmOptions.add(jvmOption);
+            }
+        }
+        return substitutedJvmOptions;
+    }
+
```
Comment (probakowski) [inline distribution/tools/launchers/src/main/java/org/elasticsearch/tools/launchers/JvmOptionsParser.java]: This method can be simplified with use of `mathce.appendReplacement` method:
```java
private static final Pattern SUBSTITUTION_PATTERN = Pattern.compile("\\$\\{([^}]+)}");

static List<String> substitutePlaceholders(final List<String> jvmOptions, Map<String, String> substitutions) {
    return jvmOptions.stream().map(jvmOption -> {
        StringBuilder sb = new StringBuilder();
        Matcher matcher = SUBSTITUTION_PATTERN.matcher(jvmOption);
        while (matcher.find()) {
            matcher.appendReplacement(sb, substitutions.get(matcher.group(1)));
        }
        matcher.appendTail(sb);
        return sb.toString();
    }).collect(Collectors.toList());
}
```
Or made faster (~2x) when you replace regex with ```indexOf```:
```java
static List<String> substitutePlaceholders(final List<String> jvmOptions, Map<String, String> substitutions) {
    final Map<String, String> placeholderSubstitutions =
            substitutions.entrySet().stream().collect(Collectors.toMap(e -> "${" + e.getKey() + "}", Map.Entry::getValue));
    return jvmOptions.stream().map(jvmOption -> {
        String actualJvmOption = jvmOption;
        int start = jvmOption.indexOf("${");
        if (start >= 0 && jvmOption.indexOf('}', start) > 0) {
            for (final Map.Entry<String, String> placeholderSubstitution : placeholderSubstitutions.entrySet()) {
                actualJvmOption = actualJvmOption.replace(placeholderSubstitution.getKey(), placeholderSubstitution.getValue());
            }
        }
        return actualJvmOption;
    }).collect(Collectors.toList());
}
```

**Review by rjernst — APPROVED** [review]
LGTM. I do think @przemek-grzedzielski 's alternative to use indexOf is simpler.

File: distribution/tools/launchers/src/main/java/org/elasticsearch/tools/launchers/JvmOptionsParser.java (hunk 1 above)
Comment (jasontedor) [reply distribution/tools/launchers/src/main/java/org/elasticsearch/tools/launchers/JvmOptionsParser.java]: Thanks for the suggestion @probakowski, I pushed it.

**Pull request merged**